ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java

```diff
@@ -214,14 +214,51 @@ protected int getPartitions(final KafkaTopicClient kafkaTopicClient) {
   @Override
   void validateKeyPresent(final SourceName sinkName, final Projection projection) {
 
-    final boolean atLeastOneKey = joinKey.getAllViableKeys(schema).stream()
-        .anyMatch(projection::containsExpression);
+    if (joinKey.isForeignKey()) {
+      final DataSource leftInputTable = getLeftmostSourceNode().getDataSource();
+      final SourceName leftInputTableName = leftInputTable.getName();
+      final String plainInputTableName = leftInputTableName.text();
+      final List<Column> leftInputTableKeys = leftInputTable.getSchema().key();
+      final List<Column> missingKeys =
+          leftInputTableKeys.stream().filter(
+              k -> !projection.containsExpression(
+                    new QualifiedColumnReferenceExp(leftInputTableName, k.name()))
+                  && !projection.containsExpression(new UnqualifiedColumnReferenceExp(
```

Comment:
Why do we need to check for both the original column name and the rewritten column name here, in contrast to the validation for the non-FK-join case below that is able to directly call `projection.containsExpression()`? (I assume there's a good reason -- I just don't understand what it is yet.)

Reply:
IIRC, the issue is that we cannot rely on `containsExpression()` as it just returns true/false, while we actually want to compile a list if missing key column. Happy to improve this code as a follow up if possible.

As a matter of fact, I don't like this code myself, but tried to get something to just work.

Reply:
This piece of code is inside a call to `Stream::filter()`. Isn't a helper method that returns a boolean exactly what we want?

Agreed that we can improve this in a follow-up. I just didn't understand what was different about checking for columns within the projection for PK and FK joins, and wanted to understand to have greater confidence that we haven't introduced any bugs around how FK joins are renaming columns in different situations.

Reply:
> Isn't a helper method that returns a boolean exactly what we want?

Well, yes and no. For base functionally "yes". For a good error message "no" -> if we only get a `true/false` answer, we don't know _which_ key columns are missing.

Reply:
Aren't we already collecting a list of `missingKeys`? My question is about updating lines 222-229 to become something like
```
final List<Column> missingKeys =
          leftInputTableKeys.stream().filter(
                   k -> !projection.containsExpression( ... )
          ).collect(Collectors.toList());
```
and then we can use the collected list of missingKeys in the error message, the same way the PR currently is. My question is just about simplifying the boolean filter to be a single call to `projection.containsExpression()` rather than two. Is this not possible?

Reply:
Ah. I remember now... `leftInputTableKeys` is `List<Column>` while we would need `List<Expression>` to apply `projection.containsExpression()`